File: lib/ts/src/UApiSchema.ts
import { createUApiSchemaFromFileJsonMap } from './internal/schema/CreateUApiSchemaFromFileJsonMap';
import { getSchemaFileMap } from './internal/schema/GetSchemaFileMap';
import { UFieldDeclaration } from './internal/types/UFieldDeclaration';
import { UType } from './internal/types/UType';
import { FsModule, PathModule } from './fileSystem';

export class UApiSchema {
    /**
     * A parsed uAPI schema.
     */

    original: any[];
    full: any[];
    parsed: Record<string, UType>;
    parsedRequestHeaders: Record<string, UFieldDeclaration>;
    parsedResponseHeaders: Record<string, UFieldDeclaration>;

    constructor(
        original: any[],
        full: any[],
        parsed: Record<string, UType>,
        parsedRequestHeaders: Record<string, UFieldDeclaration>,
        parsedResponseHeaders: Record<string, UFieldDeclaration>,
    ) {
        this.original = original;
        this.full = full;
        this.parsed = parsed;
        this.parsedRequestHeaders = parsedRequestHeaders;
        this.parsedResponseHeaders = parsedResponseHeaders;
    }

    static fromJson(json: string): UApiSchema {
        return createUApiSchemaFromFileJsonMap({ auto_: json });
    }

    static fromFileJsonMap(fileJsonMap: Record<string, string>): UApiSchema {
        return createUApiSchemaFromFileJsonMap(fileJsonMap);
    }

    /**
     * @param fs - node fs
     * @param path - node path
     */
    static fromDirectory(directory: string, fs: FsModule, path: PathModule): UApiSchema {
        // TODO
        const m = getSchemaFileMap(directory, fs, path);
        return createUApiSchemaFromFileJsonMap(m);
    }
}

File: lib/java/src/main/java/uapi/UApiSchema.java
package uapi;

import static uapi.internal.schema.CreateUApiSchemaFromFileJsonMap.createUApiSchemaFromFileJsonMap;
import static uapi.internal.schema.GetSchemaFileMap.getSchemaFileMap;

import java.util.List;
import java.util.Map;

import uapi.internal.types.UFieldDeclaration;
import uapi.internal.types.UType;

/**
 * A parsed uAPI schema.
 */
public class UApiSchema {

    public final List<Object> original;
    public final Map<String, UType> parsed;
    public final Map<String, UFieldDeclaration> parsedRequestHeaders;
    public final Map<String, UFieldDeclaration> parsedResponseHeaders;

    public UApiSchema(List<Object> original,
            Map<String, UType> parsed,
            Map<String, UFieldDeclaration> parsedRequestHeaders,
            Map<String, UFieldDeclaration> parsedResponseHeaders) {
        this.original = original;
        this.parsed = parsed;
        this.parsedRequestHeaders = parsedRequestHeaders;
        this.parsedResponseHeaders = parsedResponseHeaders;
    }

    public static UApiSchema fromJson(String json) {
        return createUApiSchemaFromFileJsonMap(Map.of("auto_", json));
    }

    public static UApiSchema fromFileJsonMap(Map<String, String> fileJsonMap) {
        return createUApiSchemaFromFileJsonMap(fileJsonMap);
    }

    public static UApiSchema fromDirectory(String directory) {
        final var map = getSchemaFileMap(directory);
        return createUApiSchemaFromFileJsonMap(map);
    }
}
